crates/ra_assists/src/move_bounds.rs

```diff
@@ -0,0 +1,120 @@
+use hir::db::HirDatabase;
+use ra_syntax::{
+    ast::{self, AstNode, NameOwner, TypeBoundsOwner},
+    SyntaxElement,
+    SyntaxKind::*,
+    TextRange,
+};
+
+use crate::{ast_editor::AstBuilder, Assist, AssistCtx, AssistId};
+
+pub(crate) fn move_bounds_to_where_clause(mut ctx: AssistCtx<impl HirDatabase>) -> Option<Assist> {
+    let type_param_list = ctx.node_at_offset::<ast::TypeParamList>()?;
+
+    let mut type_params = type_param_list.type_params();
+    if type_params.all(|p| p.type_bound_list().is_none()) {
+        return None;
+    }
+
+    let parent = type_param_list.syntax().parent()?;
+    if parent.children_with_tokens().find(|it| it.kind() == WHERE_CLAUSE).is_some() {
+        return None;
+    }
+
+    let anchor: SyntaxElement = match parent.kind() {
+        FN_DEF => ast::FnDef::cast(parent)?.body()?.syntax().clone().into(),
+        TRAIT_DEF => ast::TraitDef::cast(parent)?.item_list()?.syntax().clone().into(),
+        IMPL_BLOCK => ast::ImplBlock::cast(parent)?.item_list()?.syntax().clone().into(),
+        ENUM_DEF => ast::EnumDef::cast(parent)?.variant_list()?.syntax().clone().into(),
+        STRUCT_DEF => parent
+            .children_with_tokens()
+            .find(|it| it.kind() == NAMED_FIELD_DEF_LIST || it.kind() == SEMI)?,
+        _ => return None,
+    };
+
+    ctx.add_action(
+        AssistId("move_bounds_to_where_clause"),
+        "move_bounds_to_where_clause",
+        |edit| {
+            let params_with_bounds = type_param_list
+                .type_params()
+                .filter(|p| p.type_bound_list().is_some())
+                .collect::<Vec<_>>();
+
+            for param in &params_with_bounds {
+                let colon =
+                    param.syntax().children_with_tokens().find(|it| it.kind() == COLON).unwrap();
+                let start = colon.text_range().start();
+                let end = param.type_bound_list().unwrap().syntax().text_range().end();
```

Comment:
This might fire if there's a param without a bound?

Reply:
No, because we iterate the parameters with a bound only.

Reply:
Ah, riiight....

Let's add a test for this though: in the existing tests, all parameters have bounds.